compiler/src/parser.mly

```diff
@@ -244,18 +245,51 @@ prim:
 | ALIGNED { `Aligned }
 | UNALIGNED { `Unaligned }
 
+%inline access_type:
+ | c=COLON? ct=loc(utype) { c, ct }
+
 %inline mem_access:
-| ct=parens(utype)? LBRACKET al=unaligned? v=var e=mem_ofs? RBRACKET
-  { al, ct, v, e }
+| ct=loc(parens(utype)) LBRACKET al=unaligned? v=var e=mem_ofs? RBRACKET
+  { let s = Wsize.string_of_wsize (L.unloc ct) in
+    Utils.warning Deprecated (Location.of_loc ct)
+       "Syntax (u%s)[x + e] is deprecated. Use [:u%s x + e] instead" s s ;
+    al, Some (L.unloc ct), v, e }
+| LBRACKET al=unaligned? ct=access_type? v=var e=mem_ofs? RBRACKET
+  {
+    let ct =
+      match ct with
+      | Some (c, ct) ->
+        if c = None then begin
+          let s = Wsize.string_of_wsize (L.unloc ct) in
+          Utils.warning Deprecated (Location.of_loc ct)
+             "Syntax [u%s x + e] is deprecated. Use [:u%s x + e] instead" s s
+        end;
```

Comment:
This syntax was not allowed before, and we now allow it with a deprecated warning. Is this really what we want? I'd find it simpler to just reject this syntax.

Reply:
You are right, I will fix this.

Reply:
In fact in it maybe more complex to do than expected (some code is shared with array). 
You can discuss of this tomorrow.

Reply:
I think you can keep the structure as is, but just fail instead of issuing a warning. I mean, it can be a valid input, but on this input, you fail.

Reply:
Yes you are write